subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -244,11 +205,28 @@ private void processTransitiveDependentClasses(Set<String> visitedClasses,
             }
             accessibleResultClasses.add(accessibleDependentClass);
             DependentsSet currentDependents = getDependents(accessibleDependentClass);
-            if (!currentDependents.isDependencyToAll()) {
+            if (currentDependents.isDependencyToAll()) {
+                return currentDependents.getDescription();
```

Comment:
These full recompile reasons were lost before, which meant that things like changing a constant used by `module-info` did not lead to the expected full recompile. Instead, special-case code for `module-info` had been added further down. With this fix, that code could now be removed. This probably also fixes other issues, since `module-info` is not the only full recompile reason, just the most prominent one.